### PR TITLE
refactor: add the new Deployer API and implement K8s deployer to replace the old manager

### PR DESCRIPTION
**Pull request by zyy17**

### CODE REVIEW

File: pkg/cmd/gtctl/cluster/create/create.go [hunk 1]
```diff
@@ -86,45 +80,40 @@ func NewCreateClusterCommand(l logger.Logger) *cobra.Command {
 			l.V(0).Infof("Creating GreptimeDB cluster '%s' in namespace '%s' ...", logger.Bold(clusterName), logger.Bold(options.Namespace))
 
 			spinner.Start("Installing greptimedb-operator...")
-			if err := m.CreateOperator(ctx, createOperatorOptions); err != nil {
+			createGreptimeDBOperatorOptions := &deployer.CreateGreptimeDBOperatorOptions{
+				GreptimeDBOperatorChartVersion: options.GreptimeDBOperatorChartVersion,
+				ImageRegistry:                  options.ImageRegistry,
+			}
+			name := types.NamespacedName{Namespace: options.Namespace, Name: "greptimedb-operator"}.String()
```
Comment (daviderli614) [inline pkg/cmd/gtctl/cluster/create/create.go:87]: The name can be used a parameter, define 'options.Name'.

File: pkg/cmd/gtctl/cluster/delete/delete.go [hunk 2]
```diff
@@ -46,41 +48,40 @@ func NewDeleteClusterCommand(l logger.Logger) *cobra.Command {
...
-				Namespace:   options.Namespace,
-			})
+			name := types.NamespacedName{Namespace: options.Namespace, Name: clusterName}.String()
+			cluster, err := k8sDeployer.GetGreptimeDBCluster(ctx, name, nil)
 			if errors.IsNotFound(err) {
 				l.V(0).Infof("Cluster '%s' in '%s' not found\n", clusterName, namespace)
 				return nil
 			}
-			if err != nil && !errors.IsNotFound(err) {
+			if err != nil {
```
Comment (daviderli614) [inline pkg/cmd/gtctl/cluster/delete/delete.go:63]: It's look good to me.

File: pkg/cmd/gtctl/cluster/list/list.go [hunk 3]
```diff
@@ -30,24 +32,29 @@ func NewListClustersCommand(l logger.Logger) *cobra.Command {
...
 			}
 
 			ctx := context.TODO()
-			clusters, err := m.ListClusters(ctx, &manager.ListClusterOptions{})
+			clusters, err := k8sDeployer.ListGreptimeDBClusters(ctx, nil)
 			if err != nil && !errors.IsNotFound(err) {
 				return err
 			}
-			if errors.IsNotFound(err) || (clusters != nil && len(clusters.Items) == 0) {
+			if errors.IsNotFound(err) || (clusters != nil && len(clusters) == 0) {
```
Comment (daviderli614) [inline pkg/cmd/gtctl/cluster/list/list.go:45]: Maybe is `if errors.IsNotFound(err) || (clusters != nil && len(clusters.Item) == 0)` ?

Comment (daviderli614) [reply pkg/cmd/gtctl/cluster/list/list.go:45]: Here can:
```
if errors.IsNotFound(err) || (clusters != nil && len(clusters) == 0) {
  l.Error("clusters not found\n")
  return nil
}

clusters, err := k8sDeployer.ListGreptimeDBClusters(ctx, nil)
	if err != nil  { // Delete check [!errors.IsNotFound(err)].
        return err
}
```

File: pkg/deployer/k8s/deployer.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	if err := d.client.UpdateCluster(ctx, resourceNamespace, newCluster); err != nil {
+		return err
+	}
+
+	return d.client.WaitForClusterReady(resourceName, resourceNamespace, d.timeout)
+}
+
+func (d *deployer) DeleteGreptimeDBCluster(ctx context.Context, name string, options *DeleteGreptimeDBClusterOption) error {
```
Comment (daviderli614) [inline pkg/deployer/k8s/deployer.go:168]: The name parameter is namespaceName?  We can change to namespaceName ?

File: pkg/deployer/k8s/deployer.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+	if err != nil {
+		return err
+	}
+
+	chart, err := d.render.LoadChartFromRemoteCharts(downloadURL)
+	if err != nil {
+		return err
+	}
+
+	manifests, err := d.render.GenerateManifests(resourceName, resourceNamespace, chart, values)
```
Comment (daviderli614) [inline pkg/deployer/k8s/deployer.go:197]: GenerateManifests parameter resourceName and resourceNamespace, change it's use name and namespace is enough.

File: pkg/deployer/types.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+	// ListGreptimeDBClusters list all the deployed GreptimeDBClusters.
+	ListGreptimeDBClusters(ctx context.Context, options *ListGreptimeDBClustersOptions) ([]*GreptimeDBCluster, error)
+
+	// CreateGreptimeDBCluster create a GreptimeDBCluster with the given cluster name.
+	// The name is the namespaced name(<namespace>/<name>) in Kubernetes.
+	CreateGreptimeDBCluster(ctx context.Context, name string, options *CreateGreptimeDBClusterOptions) error
+
+	// UpdateGreptimeDBCluster update the GreptimeDBCluster spec with the given cluster name.
+	// The name is the namespaced name(<namespace>/<name>) in Kubernetes.
```
Comment (daviderli614) [inline pkg/deployer/types.go:35]: It is better to use namespaceName ？

File: pkg/deployer/k8s/deployer.go (hunk 4 above)
Comment (zyy17) [reply pkg/deployer/k8s/deployer.go:168]: For K8s deployer, it's namespacedName.

File: pkg/deployer/types.go (hunk 6 above)
Comment (zyy17) [reply pkg/deployer/types.go:35]: Let's keep API more general. The API of the deployer can run not only on K8s but also on the bare-metal environment. For that reason, `name` would be the better naming.

File: pkg/deployer/k8s/deployer.go (hunk 5 above)
Comment (zyy17) [reply pkg/deployer/k8s/deployer.go:197]: It's nothing different from using shorter variable names.

File: pkg/cmd/gtctl/cluster/delete/delete.go (hunk 2 above)
Comment (zyy17) [reply pkg/cmd/gtctl/cluster/delete/delete.go:63]: You can leave the error handling as the next PR.

File: pkg/cmd/gtctl/cluster/list/list.go (hunk 3 above)
Comment (zyy17) [reply pkg/cmd/gtctl/cluster/list/list.go:45]: As the above comment says, you can provide another PR to handle that.

File: pkg/helm/helm_test.go [hunk 7]
```diff
@@ -1,10 +1,29 @@
+// Copyright 2023 Greptime Team
```
Comment (daviderli614) [inline pkg/helm/helm_test.go:1]: Note: The year is 2023 here. In `pkg/deployer/k8s/deployer.go` and `pkg/deployer/types.go` is 2022.

Comment (zyy17) [reply pkg/helm/helm_test.go:1]: @daviderli614 we can use `license-eye` to update all the copyright in next PR.

**Review by daviderli614 — APPROVED** [review]
LGTM

**Pull request merged**